### PR TITLE
8325494: C2: Broken graph after not skipping CastII node anymore for Assertion Predicates after JDK-8309902

### PR DESCRIPTION
**Pull request by rwestrel**
After range check elimination, a cast in the main loop becomes top
because the type of its input (that depends on the iv phi) and the
type recorded in the cast do not intersect. This is a case that's
expected to be caught by assert predicates but, in this particular
case, no assert predicate constant folds.

The stride for the loop is -2. The iv phi type is `min+1..0`

As a consequence, the init value for the main loop has type int.

The range check that causes the issue is for array access:
```
lArrFld[i11 + 1] = 6;
```

The main loop is unrolled once. The second access in the loop is at
`i11 - 1` which has type `min..-1`. The range check cast at that
access becomes top. The assert predicates operates on an init value
that has the shape:

```
(CastII (AddI pre_loop_iv -2) int)
```

and type int.

That `CastII` is inserted by `PhaseIdealLoop::cast_incr_before_loop()`.

The assert predicate for the first iteration in the main loop is for
index:

```
(AddI (CastII (AddI pre_loop_iv -2) int) 1)
```

And for the second:

```
(AddI (CastII (AddI pre_loop_iv -2) int) -1)
```

Both have type int so the assert predicate can't constant fold.

I initially fixed this by changing the type of the cast from int to
the type of the iv phi:

```
(AddI (CastII (AddI pre_loop_iv -2) min+1..0) -1)
```

That allows the assert predicate for the second iteration to constant
fold. But I was then worried narrowing the type of the cast would
causes issues going forward so instead, I propose proceeding as in
8282592 and have assert predicates skip over the CastII (that part of
8282592 was later undone):

```
(AddI (AddI pre_loop_iv -2) 1)
```

which allows the assert predicate for the first iteration in the main
loop to constant fold.

The change from 8282592 caused issues because we used to narrow the
type of a cast based on the condition that guards it. That was removed
by 8319372.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8325494](https://bugs.openjdk.org/browse/JDK-8325494): C2: Broken graph after not skipping CastII node anymore for Assertion Predicates after JDK-8309902 (**Bug** - P3)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18724/head:pull/18724` \
`$ git checkout pull/18724`

Update a local copy of the PR: \
`$ git checkout pull/18724` \
`$ git pull https://git.openjdk.org/jdk.git pull/18724/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18724`

View PR using the GUI difftool: \
`$ git pr show -t 18724`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18724.diff">https://git.openjdk.org/jdk/pull/18724.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18724#issuecomment-2047595088)

### CODE REVIEW

File: src/hotspot/share/opto/loopTransform.cpp [hunk 1]
```diff
@@ -2013,6 +2013,12 @@ bool IdealLoopTree::is_invariant(Node* n) const {
 // to the new stride.
 void PhaseIdealLoop::update_main_loop_assertion_predicates(Node* ctrl, CountedLoopNode* loop_head, Node* init,
                                                            const int stride_con) {
+  if (init->Opcode() == Op_CastII) {
+    // skip over the cast added by PhaseIdealLoop::cast_incr_before_loop() when pre/post/main loops are created because
+    // it can get in the way of type propagation
+    assert(((CastIINode*)init)->carry_dependency() && loop_head->skip_assertion_predicates_with_halt() == init->in(0), "casted iv phi from pre loop expected");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopTransform.cpp:2019]: You can use `is_CastII()` and `as_CastII()`.

File: test/hotspot/jtreg/compiler/loopopts/TestAssertPredicateDoesntConstantFold.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+/*
+ * @test
+ * @bug 8325494
+ * @summary C2: Broken graph after not skipping CastII node anymore for Assertion Predicates after JDK-8309902
+ * @run main/othervm -XX:-TieredCompilation -Xcomp -XX:CompileOnly=TestAssertPredicateDoesntConstantFold::test TestAssertPredicateDoesntConstantFold
+ *
+ */
+
+public class TestAssertPredicateDoesntConstantFold {
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/loopopts/TestAssertPredicateDoesntConstantFold.java:32]: I suggest to move this test to the other predicate tests in `compiler/predicates`

Nit:
```suggestion
public class TestAssertionPredicateDoesntConstantFold {
```

**Review by chhagedorn — APPROVED** [review]
I agree with that. I thought about same now that JDK-8319372 is in.

Two minor comments, otherwise, looks good!

**Comment by rwestrel** [comment]
> I agree with that. I thought about same now that JDK-8319372 is in.
> 
> Two minor comments, otherwise, looks good!

Thanks for reviewing this. I pushed an update.

**Review by eme64 — APPROVED** [review]
Looks reasonable.

**Comment by rwestrel** [comment]
@chhagedorn @eme64 thanks for the reviews

**Pull request closed without merging**